sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,178 @@
+
+    Example
+    -------
+    >>> # X has two examples, with X[:, 2] as categorical
+    >>> X = [[0, 1, 0, 4], \
+             [6, 7, 1, 5]]
+    >>> from sklearn.preprocessing import Discretizer
+    >>> discretizer = Discretizer(n_bins=3, categorical_features=[2])
+    >>> discretizer.fit(X) # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    Discretizer(...)
+    >>> discretizer.cut_points_ # doctest: +ELLIPSIS, +NORMALIZE_WHITESPACE
+    array([[ 2. , 3. , 4.3333...],
+           [ 4. , 5. , 4.6666...]])
+    >>> # Transforming X will move the categorical features to the last indices
+    >>> discretizer.transform(X)
+    array([[0, 0, 0, 0],
+           [2, 2, 2, 1]])
+    """
+    sparse_formats = ['csr', 'csc']
+
+    def __init__(self, n_bins=2, categorical_features="all"):
+        if n_bins < 2:
+            raise ValueError("Discretizer received an invalid number "
+                             "of bins")
+        self.n_bins = n_bins
+        self.categorical_features = categorical_features
+
+        # Attributes
+        self.min_ = None
+        self.max_ = None
+        self.cut_points_ = None
+        self.n_features_ = None
+        self.continuous_mask_ = None
+
+    def _check_sparse(self, X, ravel=True):
+        if ravel:
+            return X.toarray().ravel() if sp.issparse(X) else X
+        return X.toarray() if sp.issparse(X) else X
+
+    def _set_continuous_mask(self):
+        """Sets a boolean array that determines which columns are
+        continuous features.
+        """
+        if self.categorical_features == "all":
+            self.n_continuous_features_ = self.n_features_
+            self.continuous_mask_ = np.ones(self.n_features_, dtype=bool)
+            return
+
+        if len(self.categorical_features) > self.n_features_:
+            raise ValueError("The number of categorical indices is more than "
+                             "the number of features in the dataset.")
+
+        self.continuous_mask_ = np.ones(self.n_features_, dtype=bool)
+        self.continuous_mask_[self.categorical_features] = False
+
+        # Checks if there are duplicate columns from self.categorical_features,
+        # such as [-1, 0, 0, self.n_features - 1]
+        if self.n_features_ - len(self.categorical_features) \
+                != self.continuous_mask_.sum():
+            raise ValueError("Duplicate indices detected from input "
+                             "categorical indices.")
+
+    def fit(self, X, y=None):
+        """Finds the intervals of interest from the input data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse-matrix}, shape [n_samples, n_features]
+            The input data containing continuous features to be
+            discretized. Categorical columns are indicated by
+            categorical_columns from the constructor.
+
+        Returns
+        -------
+        self
+        """
+
+        X = check_array(X, accept_sparse=self.sparse_formats)
```

Comment:
CSR is fine if categorical_features="all" but you'll probably need CSC if not because you need feature-wise access to the data. Have a look at the `data`, `indices`, `indptr` attributes of the CSC format. You'll need to access those to filter out continuous features. If categorical_features="all", you will only need `data` so both CSR and CSC should be fine.

BTW, ideally the transformation should preserve sparsity. So zeros should remain zeros.